### PR TITLE
Added print statements for probabilities without .max(axis=1), made various changes in training script, evaluation script and data processing script.

### PR DESCRIPTION
**Pull request by AnnikaSimonsen**

### CODE REVIEW

File: src/european_values/data_processing.py [hunk 1]
```diff
@@ -15,17 +15,21 @@
...
+) -> tuple[pd.DataFrame, MinMaxScaler]:
     """Process the survey data.
 
     Args:
         df:
             The survey data.
         config:
             The Hydra config.
+        normalize:
+            Whether to apply normalization. If False, scaler is fitted but not applied.
```
Comment (saattrupdan) [inline src/european_values/data_processing.py:29]:
```suggestion
        normalize (optional):
            Whether to apply normalization. If False, scaler is fitted but not applied.
            Defaults to True.
```

File: src/european_values/generative_training.py [hunk 2]
```diff
@@ -19,7 +22,7 @@ def train_generative_model(
     """Train a Gaussian Mixture Model on EU survey data.
 
     Args:
-        survey_df: Survey data DataFrame
+        survey_df: Survey data DataFrame (already processed and subset-filtered)
```
Comment (saattrupdan) [inline src/european_values/generative_training.py:25]: Adding some clarification about non-normalisation. Removed the "processed" to avoid confusion with whether processing includes normalisation.

```suggestion
        survey_df: A non-normalised dataframe containing the survey responses.
```

File: src/european_values/generative_training.py [hunk 3]
```diff
@@ -33,7 +36,7 @@ def train_generative_model(
         f"{eu_df['country_code'].nunique()} EU countries"
     )
 
-    # Get question columns
+    # Get question columns - these are already processed and normalized
```
Comment (saattrupdan) [inline src/european_values/generative_training.py:39]: Not normalised!

```suggestion
    # Get question columns
```

File: src/european_values/generative_training.py [hunk 4]
```diff
@@ -54,6 +57,7 @@ def train_generative_model(
         val_dfs.append(country_data.iloc[n_test : n_test + n_val])
         train_dfs.append(country_data.iloc[n_test + n_val :])
 
+    # Get the processed data (already normalized from process_data)
```
Comment (saattrupdan) [inline src/european_values/generative_training.py:60]: Not normalised!

```suggestion
    # Set up the data as NumPy arrays
```

File: src/european_values/generative_training.py [hunk 5]
```diff
@@ -68,39 +72,82 @@ def train_generative_model(
...
-    # Create GMM instance once
-    gmm = GaussianMixture(random_state=seed)
-
     bic_scores = {}
+
     for n_comp in n_components_range:
-        gmm.n_components = n_comp
-        gmm.fit(train_matrix)
-        bic_scores[n_comp] = gmm.bic(val_matrix)
+        # Create pipeline with scaler + GMM (as requested)
```
Comment (saattrupdan) [inline src/european_values/generative_training.py:78]: Requested?

```suggestion
        # Create pipeline with scaler + GMM
```

File: src/european_values/generative_training.py [hunk 6]
```diff
@@ -68,39 +72,82 @@ def train_generative_model(
     n_components_range = [1, 2, 3, 4, 5, 10, 15, 20, 25, 30]
     n_components_range = [n for n in n_components_range if n <= max_components]
 
-    # Create GMM instance once
-    gmm = GaussianMixture(random_state=seed)
-
     bic_scores = {}
+
```
Comment (saattrupdan) [inline src/european_values/generative_training.py:76]: Nit: Removing the newline indicates that the `bic_scores = {}` line belongs to the code block afterwards.

```suggestion
```

File: src/european_values/generative_training.py [hunk 7]
```diff
@@ -68,39 +72,82 @@ def train_generative_model(
...
-    model_dir = Path("models")
-    model_dir.mkdir(exist_ok=True)
+    logger.info("Testing probabilities on training data...")
+    train_probabilities = gmm.predict_proba(scaled_full_matrix)
+    print(f"Training data - Component weights: {gmm.weights_}")
+    print(f"Training data - First 5 full probabilities:\n{train_probabilities[:5]}")
+
+    # Try weighted approach
+    weighted_probs = (
+        np.sum(train_probabilities * gmm.weights_, axis=1) / gmm.n_components
```
Comment (saattrupdan) [inline src/european_values/generative_training.py:128]: Do we know that the multiplication is correct here? Since `gmm.weights_` is (I think) an array of shape `(n_components,)` and `train_probabilities` is of shape `(n_samples, n_components)`. We could, to be completely sure, replace `gmm.weights_` with `np.expand_dims(gmm.weights_, axis=0)`.

Also, thinking about it, we actually don't need to divide by number of components, since the weights sum up to 1.

```suggestion
        np.sum(train_probabilities * np.expand_dims(gmm.weights_, axis=0), axis=1)
```

File: src/european_values/llm_evaluation.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+import joblib
+import numpy as np
+import pandas as pd
+from sklearn.pipeline import Pipeline
+
+
+def load_gmm_pipeline(model_path: str) -> Pipeline:
+    """Load trained GMM pipeline (scaler + model)."""
+    return joblib.load(model_path)
```
Comment (saattrupdan) [inline src/european_values/llm_evaluation.py:13]: Nit: I feel like this function isn't really needed, since just writing out the code is just as long?

File: src/european_values/llm_evaluation.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+
+def process_responses(responses: np.ndarray) -> np.ndarray:
```
Comment (saattrupdan) [inline src/european_values/llm_evaluation.py:16]: Is this function actually necessary? Shouldn't all NaN's already be imputed in the data?

File: src/european_values/llm_evaluation.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    responses: np.ndarray, gmm_pipeline: Pipeline
+) -> Tuple[float, np.ndarray]:
+    """Evaluate responses using GMM pipeline."""
+    # Get GMM component to use predict_proba
+    gmm_model = gmm_pipeline.named_steps["gaussianmixture"]
+
+    # Transform data through the pipeline's scaler
+    scaled_responses = gmm_pipeline.named_steps["minmaxscaler"].transform(responses)
+
+    full_probabilities = gmm_model.predict_proba(scaled_responses)
```
Comment (saattrupdan) [inline src/european_values/llm_evaluation.py:42]: Is it not possible to simply run `gmm_pipeline.predict_proba(responses)`, or does that throw an error?

File: src/scripts/evaluate_llm_benchmark.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+logger = logging.getLogger("evaluate_llm")
+
+
+@hydra.main(config_path="../../config", config_name="config", version_base=None)
+def main(config: DictConfig) -> None:
+    """Main evaluation function."""
+    # Load data
+    logger.info("Loading data...")
+    df = load_evs_wvs_data()
```
Comment (saattrupdan) [inline src/scripts/evaluate_llm_benchmark.py:21]: Can you replace this with the data loading block from the other scripts? That allows it to be used with the trends dataset as well.

File: src/scripts/evaluate_llm_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+    # Process data but SKIP normalization (let pipeline handle it)
+    df, _ = process_data(df=df, config=config, normalize=False)  # Fixed!
```
Comment (saattrupdan) [inline src/scripts/evaluate_llm_benchmark.py:24]:
```suggestion
    df, _ = process_data(df=df, config=config, normalize=False)
```

File: src/scripts/evaluate_llm_benchmark.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+        question_cols_to_remove = [
+            col
+            for col in df.columns
+            if col.startswith("question_") and col not in question_subset
+        ]
+        df.drop(columns=question_cols_to_remove, inplace=True)
+        logger.info(f"Using {len(question_subset)} questions from subset")
+
+    # Set evaluation parameters
+    region = getattr(config.evaluation, "region", "EU")
```
Comment (saattrupdan) [inline src/scripts/evaluate_llm_benchmark.py:43]: We can assume that `region` is already set in the config. If not, it should be:

```suggestion
    region = config.evaluation.region
```

File: src/scripts/evaluate_llm_benchmark.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    model_path = getattr(
+        config.evaluation,
+        "gmm_model_path",
+        "data/processed/gmm_model/gmm_n4_seed4242.pkl",
+    )
```
Comment (saattrupdan) [inline src/scripts/evaluate_llm_benchmark.py:48]: Same with `gmm_model_path` - should always be set:

```suggestion
    model_path = config.evaluation.gmm_model_path
```

File: src/scripts/train_generative_model.py [hunk 15]
```diff
@@ -19,56 +15,32 @@
...
-            logger.info("Loading only EVS/WVS data...")
-            df = load_evs_wvs_data()
-        case _:
-            raise ValueError(
-                "At least one of `include_evs_trend` or `include_evs_wvs` must be True."
-            )
+    """Main function."""
+    # Load data
+    logger.info("Loading only EVS/WVS data...")
+    df = load_evs_wvs_data()
```
Comment (saattrupdan) [inline src/scripts/train_generative_model.py:21]: Same as above, please replace with the data loading block that allows trends.

File: src/scripts/train_generative_model.py [hunk 16]
```diff
@@ -19,56 +15,32 @@
...
 
-    logger.info("Processing the data...")
-    df = process_data(df=df, config=config)
-    logger.info(f"Shape of the data after processing: {df.shape}")
+    # Process data but SKIP normalization (let pipeline handle it)
+    logger.info("Processing the data WITHOUT normalization...")
+    df, _ = process_data(df=df, config=config, normalize=False)  # Fixed!
```
Comment (saattrupdan) [inline src/scripts/train_generative_model.py:25]:
```suggestion
    df, _ = process_data(df=df, config=config, normalize=False)
```

**Review by saattrupdan — CHANGES_REQUESTED** [review]

File: src/scripts/train_generative_model.py [hunk 17]
```diff
@@ -6,9 +6,11 @@
 import pandas as pd
 from omegaconf import DictConfig
 
-from european_values.data_loading import load_evs_wvs_data
+from european_values.data_loading import load_evs_trend_data, load_evs_wvs_data
 from european_values.data_processing import process_data
-from european_values.generative_training import train_generative_model
+from european_values.generative_training import (
+    train_generative_model,  # <-- This was missing!
```
Comment (saattrupdan) [inline src/scripts/train_generative_model.py:12]: No internal comments in the code please 🙂 

```suggestion
    train_generative_model
```
